src/cljs/owlet_ui/firebase.cljs

```diff
@@ -406,36 +406,92 @@
+          snap-complete ; Key-value pair for the "complete" function expected by
+                        ; firebase.storage.UploadTask's .on method, which is a
+                        ; 0-arg function. So we wrap the given
+                        ; :complete-with-snapshot 1-arg function, passing it the
+                        ; new file URL.
+                        [:complete
+                         #(complete-with-snapshot (.-snapshot task))]]
+
+      ; Execute the file-upload task, substituting :error or :complete functions,
+      ; if necessary.
+      (.on task
+           js/firebase.storage.TaskEvent.STATE_CHANGED
+           (-> options
+               (dissoc :into-dir :complete-with-snapshot)   ; Keys not expected by .on.
+               (conj (when-not error default-error))
+               (conj (when complete-with-snapshot snap-complete))
+               clj->js)))
+
+    (let [err-msg "Please select a file to be uploaded."]
+      (if error
+        (error (js/Error. err-msg))
+        (js/console.log "upload-file:" err-msg)))))
+
+
+(defn ez-upload-file
+  "Provides a straightforward way for a GUI to upload a local file to Firebase
+  Storage and receive the URL pointing to the new file there.
+
+  When called, ez-upload-file initiates the upload of the file selected in an
+  <input type=\"file\" id=input-elem-id> element in the current document,
+  where input-elem-id is the string given in the first argument. The given
+  destination directory string must be relative to the root of Firebase
+  Storage. As the upload proceeds, atom progress-pct-atom is periodically
+  updated with a number: the percentage of uploaded bytes. Atom error-atom will
+  contain the nil value unless the user did not provide a file in the input
+  element or if there was some other problem preventing the upload. In this
+  case error-atom will contain an error message.
+
+  Finally, when the upload is complete, a re-frame event is dispatched, which
+  is a vector of the given event-id, the new URL string, and any provided args.
+  "
+  [input-elem-id dest-dir progress-pct-atom error-atom event-id & args]
+
+  (when progress-pct-atom (reset! progress-pct-atom 0))
+  (when error-atom        (reset! error-atom        nil))
+  (apply
+    upload-file
+
+    ; JS File object, or nil if none selected:
+    (-> input-elem-id js/document.getElementById (aget "files" 0))
+
+    ; Upload options:
+    (concat
+      [:into-dir               dest-dir
+       :complete-with-snapshot (fn [snapshot]
+                                 (->> snapshot          ; The firebase.storage.UploadTaskSnapshot
+                                      .-downloadURL     ; The new URL of the stored file.
+                                      (conj args)       ; Makes (url arg1 arg2 ...)
+                                      (into [event-id]) ; Makes [event-id url arg1 arg2 ...)
+                                      rf/dispatch))]
+      (and progress-pct-atom
```

Comment:
@tylerperkins how does this work ?
`(concat ... (and ...) (and ...))`
Im intrigued.